### PR TITLE
Add concurrency safeguards to the current round timeout field on chain

### LINKED ISSUE #629: Possible race condition
`Chain.crtCount` is not synchronized, so ` roundTimeoutCount != mc.GetRoundTimeoutCount()` looks awkward to check wether timeout happened, at least it should be synchronized or atomic type used

```go
func (mc *Chain) GenerateBlock(ctx context.Context, b *block.Block,
	bsh chain.BlockStateHandler, waitOver bool) error {

	var roundTimeoutCount = mc.GetRoundTimeoutCount()
	var txnIterHandler = func(ctx context.Context, qe datastore.CollectionEntity) bool {
		count++
		if mc.GetCurrentRound() > b.Round {
			roundMismatch = true
			return false
		}
		if roundTimeoutCount != mc.GetRoundTimeoutCount() {
			roundTimeout = true
			return false
		}
...

### PR DESCRIPTION
**Pull request by fabioDMFerreira**
## Fixes
Add concurrency safeguards to the current round timeout field on chain (#629)
Also, add concurrency safeguards to the transaction cost as the processing of transactions may be done concurrently.

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by dabasov** [comment]
the problem here is that tii reference escapes from generate block and most likely is used concurrently in this closure. It is set to handler and 
```go
func txnIterHandlerFunc(mc *Chain,
	b *block.Block,
	lfb *block.Block,
	bState util.MerklePatriciaTrieI,
	txnProcessor txnProcessorHandler,
	tii *TxnIterInfo) func(context.Context, datastore.CollectionEntity) bool {
	return func(ctx context.Context, qe datastore.CollectionEntity) bool {
```
and handler lately is passed to iterate collection where it can be used concurrently
```go
	err := transactionEntityMetadata.GetStore().IterateCollection(cctx, transactionEntityMetadata, collectionName, txnIterHandler)
```

please check all tii reference sharing between go routines

**Comment by fabioDMFerreira** [comment]
> the problem here is that tii reference escapes from generate block and most likely is used concurrently in this closure. It is set to handler and
> 
> ```go
> func txnIterHandlerFunc(mc *Chain,
> 	b *block.Block,
> 	lfb *block.Block,
> 	bState util.MerklePatriciaTrieI,
> 	txnProcessor txnProcessorHandler,
> 	tii *TxnIterInfo) func(context.Context, datastore.CollectionEntity) bool {
> 	return func(ctx context.Context, qe datastore.CollectionEntity) bool {
> ```
> 
> and handler lately is passed to iterate collection where it can be used concurrently
> 
> ```go
> 	err := transactionEntityMetadata.GetStore().IterateCollection(cctx, transactionEntityMetadata, collectionName, txnIterHandler)
> ```
> 
> please check all tii reference sharing between go routines

I got it. Only now, I realize tii escapes to `txnProcessor` which also updates fields. Thanks

File: code/go/0chain.net/miner/protocol_block.go [hunk 1]
```diff
@@ -726,7 +727,7 @@ type TxnIterInfo struct {
 	// included transaction data size
 	byteSize int64
 	//accumulated transaction cost
-	cost int
+	cost int64
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_block.go]: cost works well I think

**Comment by dabasov** [comment]
please fix lint error

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**